Improve cgroup device rule validation

LeviHarrison opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Fixes #41920

- Passing only `a` is now allowed
- When one passes only `a`, a cgroup rule allowing all permissions is returned
- If an `a` type rule is passed but is not `a *:* rwm` (`rwm` can be reordered), an error is returned informing the user that the rest of their rule is ineffective ([reference](https://github.com/torvalds/linux/blob/v5.10/Documentation/admin-guide/cgroup-v1/devices.rst))
- Changed the validation of `major` and `minor` numbers so that they are 12 and 20 bits respectively
- Changed the behavior of all (`*`) major or minor numbers to just omit the field in the cgroup permission, as per the [runtime-spec](https://github.com/opencontainers/runtime-spec/blob/v1.0.2/config-linux.md#device-whitelist)
- Updated `oci` tests to reflect these changes
- Moved cgroup validation to container create, instead of container start
- Changed the name of the `AppendDevicePermissionsFromCgroupRules` function to `DevicePermissionsFromCgroupRules` and changed the parameters to reflect that it doesn't need to do any appending anymore
- Added an `errdef` for an error returned by the `DevicePermissionsFromCgroupRules` function and added tests for that `errdef`

One more thing the issue calls for is either updating the CLI or removing the client-side validation. This seems like a decision that has to be made, so I decided to leave it out of this PR for now. I'd be happy to add it once an agreement has been reached.

Also, I just noticed that the regex allows rules such as `c 1:1 rrr` and `b 1:1 wmm`. Should it?

**- How I did it**

- To validate the `a` type rules, I added a function that checks the rule to see if it is either a singular `a` or, for example, a `a 1:5 w`. I then added a second regex that parses the rule to make sure that it's either `a` or `a *:* rwm` (`rwm` can be reordered), and throw an error telling the user that their rule is ineffective if the regex doesn't pass.
- Previously, the cgroup rule validation took place in the `WithDevices` function, which is called through `daemon.createSpec`, which is called on container start. `AppendDevicePermissionsFromCgroupRules` is passed a field from the config (`HostConfig.DeviceCgroupRules`), which is actually set on container create. So, I moved `AppendDevicePermissionsFromCgroupRules` to right after that field is set in container create, and set the result of it to a new field (`HostConfig.DeviceCgroupPermissions`). Because that `WithDevices` function has access to the config, I replaced the block of code that ran the function with one that just appends `HostConfig.DeviceCgroupPermissions` to the accumulated `devPermissions`. A similar pattern exists in `oci_windows.go`, and I did the same there. Because this function is now not appending to any existing value, just being used to set a field, I decided to change the name to `DevicePermissionsFromCgroupRules` and remove the parameter that provided the existing value.

**- How to verify it**

- Run `oci` tests

```shell
curl -v --unix-socket /var/run/docker.sock \
"http://localhost/v1.41/containers/create?name=foobar" \
-H "Content-Type: application/json" \
-d '{"Image":"hello-world", "HostConfig":{"DeviceCgroupRules": ["invalid rule"]}}'
```

`{"message":"invalid device cgroup rule format: 'invalid rule'"}`

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

`Improved device cgroup rule parsing and moved it to container create`

**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://user-images.githubusercontent.com/54278938/108505387-4f34a880-7285-11eb-8c21-e42ad1ac163d.png)

LeviHarrison committed:
Added singular 'a' option to device-cgroup-rule regex

As per the Linux documentation, the cgroup rule 'a' is valid and maps to 'a *:* rwm'. Previously, the regex would not accept 'a' alone.

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Actually Fixed the Regex

This time it actually passes the tests.

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Improve parsing for 'a' type cgroup rules

An 'a' type cgroup rule allows all permissions for all devices, and is effectively 'a *:* rwm', but can be expressed as just 'a'. This commit allows the singular 'a', and throws an error if an expanded rule is not 'a *:* rwm'. Tests are also updated.

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Updated Size of Cgroup Rule Major and Minor Numbers

The size cgroup rule major and minor numbers are 12 and 20 bits respectively, so I made the parsing code reflect that, and updated tests to pinpoint any inaccuracies.

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Move Cgroup Rule Validation to Container Create

Previosly, cgroup rule validation took place on container start. Now it takes place on container create. The cgroup rule validation function has also been changed to reflect that it does not do any appending anymore. Error definitions were added, and tests were updated.

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Changed '*' major and minor value

Under the runtime spec, when a major or minor value is '*', it is to be empty as to signifiy all. Tests are updated to relfect this change.

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Add nolint rule to regex

The linter was failing, and now I see why the nolint rule is on the above regex.

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Simplified Assignment

Got rid of an uneeded variable

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison commented:
Not sure what's going on with the `amd64` tests here. All the failure points work fine on my machine, and it fails at different places on different runs for seemingly not reason.

thaJeztah commented:
CI failures look unrelated and are known to be flaky (https://github.com/moby/moby/issues/36907 and https://github.com/moby/moby/issues/41561)

thaJeztah commented on oci/oci.go:
```diff
@@ -32,35 +32,47 @@ func SetCapabilities(s *specs.Spec, caplist []string) error {
 			Access: matches[4],
 		}
-		if matches[2] == "*" {
-			major := int64(-1)
```
w.r.t. Major/Minor `nil` being equivalent to `-1` (runtime-spec); Looks indeed that runc uses these as a default, and only override in the "not nil" situation; https://github.com/opencontainers/runc/blob/ee06c700dcb34504204cb171bb35e0e01477bf58/libcontainer/specconv/spec_linux.go#L475-L487

Should we check if other runtimes (e.g. `crun` handle it the same?)

thaJeztah commented on oci/oci.go:
```diff
@@ -32,35 +32,47 @@ func SetCapabilities(s *specs.Spec, caplist []string) error {
-		} else {
-			major, err := strconv.ParseInt(matches[2], 10, 64)
+		if matches[2] != "*" {
+			major, err := strconv.ParseInt(matches[2], 10, 12)
```
I'm wondering if `12 bit` is correct here. I recall the kernel has some "weird" things; e.g., using an "unsigned" integer, but allowing `-1` as "special value". In this case (just from a quick look); it looks like it's indeed using `12 bit unsigned` and `20 bit unsigned` for these; https://github.com/torvalds/linux/blob/139711f033f636cc78b6aaf7363252241b9698ef/include/linux/kdev_t.h#L7-L11

Which would mean (`-1`?) _or_ `0-4096` / `0-1048576` ?

While definitely not the "canonical source of truth", this StackOverflow question has an interesting quote as well; https://stackoverflow.com/a/14849363/1811501 (emphasis mine)

> As of Version 2.6.0 of the kernel, dev_t is a 32-bit quantity with 12 bits set aside for the major number and 20 for the minor number. ***Your code should, of course, never make any assumptions about the internal organization of device numbers; it should, instead, make use of a set of macros found in linux/kdev_t.h***

LeviHarrison replied:
Yeah, that does seem to be the case.

thaJeztah commented on oci/oci_test.go:
```diff
@@ -88,18 +88,23 @@ func TestAppendDevicePermissionsFromCgroupRules(t *testing.T) {
 			doc:         "major out of range",
-			rule:        "c 18446744073709551616:1 rwm",
-			expectedErr: `invalid major value in device cgroup rule format: 'c 18446744073709551616:1 rwm'`,
+			rule:        "c 2048:1 rwm",
```
See my other comment; we'd have to check if 2048 is indeed out of range (it may not be).

Possibly we may have to skip range validation (other than conversion to the `LinuxDeviceCgroup.Major` / `LinuxDeviceCgroup.Minor`), and leave it to the OCI runtime and the Kernel itself to decide wether or not the value is "ok"

LeviHarrison replied:
Although I think range validation is important (especially because it takes place on `create` now and the Kernel will only check on `start`, which could give a user false confidence in their rules), the general uncertainty surrounding it, with both the `runtime-spec` and `runc` ignoring the limit, gives me pause. Maybe we just return a warning, something along the lines of  `this value probably will be out of range`.

thaJeztah reviewed (changes requested):
Thanks! Had a first pass at your changes, and left comments/thoughts/suggestions inline while looking over them

LeviHarrison commented on oci/oci.go:
```diff
@@ -47,16 +47,11 @@ func DevicePermissionsFromCgroupRules(rules []string) ([]specs.LinuxDeviceCgroup
 			}
 
-			major := int64(-1)
-			minor := int64(-1)
```
`crun` does seem to [use `-1` as their default/everything value](https://github.com/containers/crun/blob/ebba51f329cd7e3b508febb1e9752b573f19f6a4/src/libcrun/cgroup.c#L2435). 

When I wrote this, I was referencing the [runtime-spec](https://github.com/opencontainers/runtime-spec/blob/v1.0.2/config-linux.md#device-whitelist), which describes that unset major/minor numbers map to the default `*`. The same goes for the Type and Access too, but that's irrelevant. For the sake of conformity and explicitness though, I'll add those defaults back.

LeviHarrison commented on daemon/oci_linux.go:
```diff
@@ -912,11 +912,7 @@ func WithDevices(daemon *Daemon, c *container.Container) coci.SpecOpts {
-			if err != nil {
-				return err
-			}
+			devPermissions = append(devPermissions, c.HostConfig.DeviceCgroupPermissions...)
```
I never thought about that aspect, thanks! I would put forth adding a migration step to avoid either adding more logic to the validation function or adding a second (less strict) one, although I can see why having those kinds of things around could be undesirable. Up to you.

LeviHarrison committed:
Changed default values for cgroup rules, change major and minor range, and added warnings

The default values for the cgroup rules have been changed back to -1, the major and minor values are now evaluated as 12 and 20 bit unsigned numbers, and a warning is returned instead of an error when there's not explicitly an error with the cgroup

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Removed cgroup rule errdef

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Added support for warnings now returned by cgroup rule validation

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Changed import formatting

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison commented:
I fixed the major and minor range, errdefs, and added warnings. I'm still waiting on further feedback before I resolve the issue with setting permissions in restarted containers.

LeviHarrison committed:
Changed cgroup rule major/minor out of range error message

The error message is now more verbose when the values are out of range

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Unified the return values in create_unix and create_windows

Previously, I had added support for returning warnings to create_unix, but not create_windows, which triggered an error on windows builds

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Captilized first word in warning

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>

LeviHarrison committed:
Changed tests to expect capital in warning

In the last commit, I capitalized the first word in the warning, but forgot to do the same in the tests, causing them to fail

Signed-off-by: Levi Harrison <levisamuelharrison@gmail.com>
